### ISSUE TITLE
SSLSocketFactory is not used for redirects in 4.4.0

### ISSUE THREAD

**Issue by Stonos** [square/okhttp#5831]
OkHttp 4.4.0
Android Studio 4.0 Beta 1
Android 9

I have configured my OkHttp client in development versions to accept all certificates:
```kotlin
private fun buildOkHttpClient(): OkHttpClient {
    val builder = OkHttpClient.Builder()

    val unsafeTrustManager = createUnsafeTrustManager()
    val sslContext = SSLContext.getInstance("SSL")
    sslContext.init(null, arrayOf(unsafeTrustManager), null)
    builder.sslSocketFactory(sslContext.socketFactory, unsafeTrustManager)

    builder.hostnameVerifier(HostnameVerifier { hostname, session -> true })

    return builder.build()
}

private fun createUnsafeTrustManager(): X509TrustManager = object : X509TrustManager {
    override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {}
    override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {}
    override fun getAcceptedIssuers(): Array<out X509Certificate>? = emptyArray()
}
```

Unfortunately, it doesn't seem to apply to HTTP 301 redirects on v4.4.0:

```kotlin
private suspend fun makeRequest(url: String) = withContext(Dispatchers.IO) {
    val okHttp = buildOkHttpClient()
    val request = Request.Builder().url(url).build()
    okHttp.newCall(request).execute()
}

makeRequest("https://bit.ly/2PwiDfE")
```

The above will fail with a SSLPeerUnverifiedException on v4.4.0 (even though the client has been configured to accept all certificates), but it will work fine with v4.3.1:

```
2020-02-28 01:56:24.632 5249-5249/com.example.okhttpcrash E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.example.okhttpcrash, PID: 5249
    javax.net.ssl.SSLPeerUnverifiedException: Failed to find a trusted cert that signed Certificate:
        Data:
            [...]
        at okhttp3.internal.tls.BasicCertificateChainCleaner.clean(BasicCertificateChainCleaner.kt:91)
        at okhttp3.internal.connection.RealConnection$connectTls$1.invoke(RealConnection.kt:393)
        at okhttp3.internal.connection.RealConnection$connectTls$1.invoke(RealConnection.kt:73)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at okhttp3.Handshake.peerCertificates(Unknown Source:7)
        at okhttp3.internal.connection.RealConnection.supportsUrl(RealConnection.kt:581)
        at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:147)
        at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:107)
        at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:75)
        at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:245)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:82)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
2020-02-28 01:56:24.634 5249-5249/com.example.okhttpcrash E/AndroidRuntime:     at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:74)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:197)
        at okhttp3.internal.connection.RealCall.execute(RealCall.kt:148)
        at com.example.okhttpcrash.MainActivity$makeRequest$2.invokeSuspend(MainActivity.kt:54)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:561)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:727)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:667)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:655)
```

However, doing the same thing with a link that doesn't redirect you works fine:
```kotlin
makeRequest("https://github.com/square/okhttp")
```

**Comment by yschimke** [comment]
Thanks for raising.  FWIW I think it's slightly different, it's not that it doesn't use the SSLSocketFactory, it's that we can avoid cleaning the certificates on the first request, but we check whether we can reuse a connection after the redirect, and this triggers the awkward situation with the Cleaner.  No sure what the fix is here.

Competing aims in the code...

1) You are trying to tell OkHttp to ignore whether the cert is valid or not.
2) OkHttp is trying to use the cert (and so wants to verify it) in order to check hostname etc.

**Comment by yschimke** [comment]
It's failing on `handshake!!.peerCertificates[0]` which triggers the cleaner.  

```
  fun supportsUrl(url: HttpUrl): Boolean {
    val routeUrl = route.address.url

    if (url.port != routeUrl.port) {
      return false // Port mismatch.
    }

    if (url.host == routeUrl.host) {
      return true // Host match. The URL is supported.
    }

    // We have a host mismatch. But if the certificate matches, we're still good.
    return !noCoalescedConnections &&
        handshake != null &&
        OkHostnameVerifier.verify(url.host, handshake!!.peerCertificates[0] as X509Certificate)
  }
```

**Comment by yschimke** [comment]
I think the fix is the same as we do here

https://github.com/square/okhttp/pull/5703/files

**Comment by yschimke** [comment]
For discussion: https://github.com/square/okhttp/pull/5835

**Comment by swankjesse** [comment]
Fixed 4.4.1.
